File: airesources/sample_bots/Python3/fastsettler.py
import math

import hlt
import logging


def fast_settler():
    my_tag, map_size, initial_map, log = yield "Fast Settler"
    turn = 0

    while True:
        turn += 1
        game_map = yield
        command_queue = []

        for ship in game_map.ships[my_tag].values():
            if ship.docked != "undocked":
                continue

            if hlt.is_warping(ship):
                data = hlt.get_warp_extra_data(ship)
                planet = game_map.planets.get(data)
                if planet and planet.owned:
                    hlt.cancel_warp(ship)
                continue

            planets = game_map.planets.values()
            sort_key = lambda planet: hlt.distance(ship, planet)
            for planet in sorted(planets, key=sort_key):
                if planet.owned:
                    if planet.owner != my_tag:
                        continue

                    if len(planet.docked_ships) >= planet.num_docking_spots:
                        continue

                    # Don't send a ship to a planet to which we've already started flying a ship
                    if any(game_map.ships[my_tag][ship_id].docked not in ("docked", "docking") for ship_id in planet.docked_ships):
                        continue

                # Prevent more than 2 from converging on the same planet
                planet.owned = True
                planet.docked_ships.append(ship.id)

                angle, distance = hlt.orient_towards(ship, planet)
                x = planet.x + int((planet.r + 1) * math.cos((angle * math.pi / 180) + math.pi))
                y = planet.y + int((planet.r + 1) * math.sin((angle * math.pi / 180) + math.pi))
                if hlt.can_dock(ship, planet):
                    logging.warn("{:03} {:02}: docking".format(turn, ship.id))
                    command_queue.append(hlt.dock(ship, planet))
                elif distance > 10 and hlt.pathable(ship, x, y):
                    logging.warn("{:03} {:02}: warping to {} {}".format(turn, ship.id, x, y))
                    hlt.warp(ship, x, y, extra_data=planet.id)
                else:
                    logging.warn("{:03} {:02}: moving to {} {}".format(turn, ship.id, planet.x, planet.y))
                    command_queue.append(hlt.move_to(ship, angle, 1))

                break

        command_queue.extend(hlt.update_warps())
        yield command_queue

hlt.run_bot(fast_settler)

File: airesources/sample_bots/C++/fastsettler.cpp
#include "../../C++/hlt/hlt.hpp"

typedef std::pair<hlt::EntityIndex, hlt::Planet> PlanetPair;

int main() {
    hlt::PlayerId my_tag;

    auto setup = hlt::initialize("FastSettler++");
    my_tag = setup.first;

    auto behaviors = hlt::BehaviorManager();

    auto moves = std::vector<hlt::Move>();
    while (true) {
        auto game_map = hlt::get_map();
        moves.clear();

        auto planets = std::vector<PlanetPair>(
            game_map.planets.begin(), game_map.planets.end());
        for (const auto& ship_pair : game_map.ships[my_tag]) {
            const auto ship_id = ship_pair.first;
            const auto& ship = ship_pair.second;

            if (behaviors.is_behaving(ship_id)) {
                continue;
            }

            if (ship.docking_status != hlt::DockingStatus::Undocked) {
                continue;
            }

            // Sort planets by distance from current ship
            std::sort(
                planets.begin(), planets.end(),
                [&](const PlanetPair& planet1, const PlanetPair& planet2) -> bool {
                    const auto distance1 = ship.location.distance(planet1.second.location);
                    const auto distance2 = ship.location.distance(planet2.second.location);
                    return distance1 < distance2;
                }
            );
            for (auto& planet_pair : planets) {
                const auto planet_id = planet_pair.first;
                auto& planet = planet_pair.second;

                if (planet.owned) {
                    continue;
                }

                // Don't let multiple ships converge on the same planet in
                // the same frame
                planet.owned = true;

                auto closest_point = game_map.closest_point(
                    ship.location, planet.location,
                    planet.radius + hlt::GameConstants::get().DOCK_RADIUS);

                const auto distance = ship.location.distance(planet.location) - planet.radius;

                if (ship.can_dock(planet)) {
                    moves.push_back(hlt::Move::dock(ship_id, planet_id));
                }
                else if (closest_point.second &&
                    game_map.pathable(ship.location, closest_point.first, 0.6) &&
                    distance > 10) {
                    behaviors.warp_to(ship_id, closest_point.first);
                }
                else {
                    const auto angle = ship.angle_to(planet);
                    const auto thrust = static_cast<unsigned short>(
                        std::min(distance, hlt::GameConstants::get().DRAG));
                    moves.push_back(hlt::Move::thrust(
                        ship_id,
                        game_map.adjust_for_collision(ship.location, angle, thrust)));
                }

                break;
            }
        }

        behaviors.update(game_map, moves);
        hlt::send_moves(moves);
    }
}
